### ISSUE TITLE
10.6.1 (previously 14.2.1) needs to be clarified to make the goal and standard to achieve clearer.

### ISSUE THREAD

**Issue by tghosth** [OWASP/ASVS#2467]
This requirement is currently not achievable at any level. We need to clarify it with what exactly we are expecting, potentially taking into account different ASVS levels.

| # | Description | L1 | L2 | L3 | CWE |
| :---: | :--- | :---: | :---: | :---: | :---: |
| **10.6.1** | [MODIFIED, MOVED FROM 14.2.1] Verify that all components are up to date. | ✓ | ✓ | ✓ | |

As a starter, I would be prepared to accept:

| # | Description | L1 | L2 | L3 | CWE |
| :---: | :--- | :---: | :---: | :---: | :---: |
| **10.6.1** | [MODIFIED, MOVED FROM 14.2.1] Verify that only component versions without exploitable vulnerabilities are in use. | ✓ | ✓ | ✓ | |

Even then, I am not sure how practically achievable that is. Potentially a requirement that you can only really test for but cannot really proactively achieve.

**Comment by elarlang** [comment]
Previously discussed in:
* https://github.com/OWASP/ASVS/issues/2164
* https://github.com/OWASP/ASVS/issues/2088

**Comment by tghosth** [comment]
I read through those and I don't think they answer my primary concern with this requirement.

**Comment by elarlang** [comment]
The word "exploitable" here is vague. If there is clear SQLi, command injection, or something like that, it is clear. But if it provides a weakness, that independently seems "not important", but chaining with other vulnerabilities it can become a critical factor for a successful attack to happen or not.

Also, I don't see a huge problem with being up-to-date. You need to have all components mapped, with the source etc. What is not achievable here?

To have components up-to-date is critical for L1. It can be more important than SQL injection in the application - because SQL injection from custom code someone needs to find separately, not up-to-date components are subject to automated mass-attack,s and the application is "pwned" without being targeted.

If you can find better wording for that, let's see that. I'm ok with the current solution and with L1.

**Comment by randomstuff** [comment]
Yes, I agree that "up to date" may not be necessary, may be vague. In some cases, having the latest v4.0 edge version might event be more vulnerable than latest v3.49.4 :) My argument are here: https://github.com/OWASP/ASVS/issues/2088#issuecomment-2355412099

**Comment by randomstuff** [comment]
I would say:

> **10.6.1** 	[MODIFIED, MOVED FROM 14.2.1] Verify that only component versions without *known* exploitable vulnerabilities are in use.

**Comment by randomstuff** [comment]
> Also, I don't see a huge problem with being up-to-date. You need to have all components mapped, with the source etc. What is not achievable here?

You are about to make a new release of your project at day D. A new version of obscure-dependency just comes out the same day you are supposed to release. Upgrading the dependency without properly testing brings a risk of introducing a regression. If the new versions does not fix security issues (and only fixes minor bugs which do not affect you) it may not be worth it.

**Comment by elarlang** [comment]
> Yes, I agree that "up to date" may not be necessary, may be vague. In some cases, having the latest v4.0 edge version might event be more vulnerable than latest v3.49.4 :) My argument are here: https://github.com/OWASP/ASVS/issues/2088#issuecomment-2355412099

Here we hit to demagogy field - if you use v3.49.4 and it is the latest v3.* you are up to date.

> 10.6.1 [MODIFIED, MOVED FROM 14.2.1] Verify that only component versions without known exploitable vulnerabilities are in use.

I already explained, why the word "exploitable" is bad and not practical. The reason to be up-to-date with everything, is that you also have fixes against the exploits that are not "known" **for you**. Probably it is not too rare that quite critical vulnerabilities are fixed silently and then you are living happily with the knowledge that your 10 releases old version do not have any known vulnerabilities.

> You are about to make a new release of your project at day D. A new version of obscure-dependency just comes out the same day you are supposed to release. Upgrading the dependency without properly testing brings a risk of introducing a regression. If the new versions does not fix security issues (and only fixes minor bugs which do not affect you) it may not be worth it.

Does this situation makes it valid to use old and _potentially_ vulnerable version? In this situation you are going to update your application when you are ready to do it and have tested it. For that, you need to have process in place, but process itself is out of ASVS scope.

If you (@tghosth and @randomstuff ) want to say, that let's leave this requirement out (even from L1), it's one of the most insane proposals I have seen in ASVS during 5 years. The rest is said here: https://github.com/OWASP/ASVS/issues/2467#issuecomment-2543916032

**Comment by jmanico** [comment]
Everyone here is right. 

It is very debatable as to wether you should keep all libraries up to date. Especially if you are a fast early mover. You might update super fast, and get hit with a purposely malicious update where someone compromises the library. It's happened plenty.

But the less you actually keep libraries up to date, the more difficult it will be to update them when you are forced to do so.

There are quite a few other tradeoffs here that folks above has discussed.

So what is the real risk? The bottom line is, we do not want libraries live that can be attacked. Period. The rest is about process, time and other secondary factors. Proving exploitability in the world of this party libraries is super expensive.

So I suggest as an imperfect middle- ground:

| # | Description | L1 | L2 | L3 | CWE |
| :---: | :--- | :---: | :---: | :---: | :---: |
10.6.1 | [MODIFIED, MOVED FROM 14.2.1] Verify that only component versions without known vulnerabilities are in use. | ✓ | ✓ | ✓

**Comment by elarlang** [comment]
> Verify that only component versions without known vulnerabilities are in use.

* is it also requirement for L3 with the same wording?
* some may also argue, that there is component with known vulnerability but it is not exploitable in my environment

**Comment by jmanico** [comment]
> is it also requirement for L3 with the same wording?

I think so, yes. There are many reasons why you may not want to use the most recent version of a library. Functionality, security, performance issues and more.
> some may also argue, that there is component with known vulnerability but it is not exploitable in my environment
This is not reasonable at scale. There is no way if I’m a Fortune 500 that I’m going to verify hundreds of thousands of library updates before making that call. It makes sense in theory and on a small scale, but breaks down for a large org. We need to trust the tools at scale.

This may change as tools get better at understanding exploitability.

**Comment by tghosth** [comment]
> But proving exploitability in the world of third-party libraries is super expensive and time consuming, and is nearly impossible at scale in a big company with thousands of libraries and updates. There is no way in hell a big corp is going to prove exploitability for each library update, they do not have enough resources.

I know. When I say "_Verify that only component versions without exploitable vulnerabilities are in use._", my intention is to put the onus on a verifier to prove that the application contains exploitable vulnerabilities. Clearly the app developer cannot verify this so the only way to prove that the app violates this requirement is for the verifier / pen tester to exploit the vulnerability.

> Verify that only component versions without known vulnerabilities are in use.

So the company fails requirement this if they use a component containing a CVE rated as low? 





I have an alternative proposal, I think maybe we need to replace this with a documentation requirement?

"_Verify that the application documentation defines a clear risk based remediation time frame for 3rd party library vulnerabilities and that these time frames are met._"

I think this would be an acceptable L1 requirement although we could think if we wanted to strengthen it for higher levels. It would not be super easy to verify but I agree that it is am important topic and the most important thing is to have a clearly documented policy which is followed.

**Comment by jmanico** [comment]
I think moving to a document based requirement is a solid idea. Very pragmatic.

**Comment by tghosth** [comment]
@elarlang @randomstuff what are your thoughts on my proposal: https://github.com/OWASP/ASVS/issues/2467#issuecomment-2550505497

**Comment by elarlang** [comment]
> @elarlang @randomstuff what are your thoughts on my proposal: [#2467 (comment)](https://github.com/OWASP/ASVS/issues/2467#issuecomment-2550505497)

From direction of "removing the requirement" is for sure step to good direction.

* it would be the first L1 documentation requirement?
* can the proposed documentation requirement be a back up solution for the case, if the application is not up to date with components? a'la "Verify that the application components are up-to-date or some-rules-from-the-risk-based-documentation."

> It is borderline a process vulnerability but I don't think there is any escaping that.

It may take process to achieve it, but we verify the result (the application), is it achieved, not how it is done (the process)

**Comment by tghosth** [comment]
> can the proposed documentation requirement be a back up solution for the case, if the application is not up to date with components? a'la "Verify that the application components are up-to-date or some-rules-from-the-risk-based-documentation."

Potentially we could word it this way

> but we verify the result (the application), is it achieved, not how it is done (the process)

Can you expand on how you would verify this?

**Comment by elarlang** [comment]
> Can you expand on how you would verify this?

I don't get the point of subgoal for the question, because - you need to have an inventory, you need to have info about repos from where components are taken, so you have information about the version you use and the source from where you can detect the latest version.

It is too obvious answer for that question so I'm not sure it was the actual point for that question.

**Comment by randomstuff** [comment]
> Verify that the application documentation […]

I'm not so fan of the indirection by "you-must-document-what-you-want-to-do-and-then-you-have-to-do-it" in general and in this case as it kind-of allows you to document crappy practice and be happy to apply the crappy practice :)

> So the company fails requirement this if they use a component containing a CVE rated as low?

I think this requirement should fail if there is a low CVE (without more information). The CVE might be low in general but actually have a real impact in this context/application.

I think it might be OK to include a dependency with a vulnerability if an analysis of the CVE in the context of this specific application is done and the result of this analysis is that the vulnerability is:

* not applicable in this context;
* or is properly mitigated in this context;
* or as a very low impact (wrt. the security requirements of this application).

But this is departing a lot from "always apply the latest update".

**Comment by jmanico** [comment]
> But this is departing a lot from "always apply the latest update".

This is very good because always applying the latest update - while often is a good idea - can also sometimes lead to security and functionality bugs. 

It’s not at all realistic in the real world to *always* use the latest versions of libraries, especially when the latest has security bugs. Which absolutely happens.

**Comment by randomstuff** [comment]
> This is very good because always applying the latest update - while often is a good idea - can also sometimes lead to security and functionality bugs.

I agree, I was *not* in favor of "always apply the latest update" :)

**Comment by jmanico** [comment]
Good! 😁

**Comment by elarlang** [comment]
For me it seems that you are so much blinded and in love with the theory "the latest release may contain a security hole", that you are losing the focus what is the actual problem to solve here.

If you could do the stats - what is most likely the most secure release? The latest one or something from the past?

---

I recommend to change the focus here: the problem to solve is "how to mitigate the risk - downgrade the likelihood and the impact - from security weaknesses from the used 3rd party components".

Clearly, there is no "one fits all" solution. Being up-to-date in my opinion is a way better solution than "for all components, analyze all versions from that are newer than you use, analyze all known CVE's from those and measure, are you potentially affected" and this all relies on hope that there is no "silent fix" for a critical vulnerability. Convince me that it is somehow cheaper, easier or more secure way to handle it than being up-to-date.

Proposed direction:
> Verify that the risk of using vulnerable 3rd party component is minimized by using up to date or latest "known to be safe" version, -add here your other options-

**Comment by tghosth** [comment]
> > Can you expand on how you would verify this?
> 
> I don't get the point of subgoal for the question, because - you need to have an inventory, you need to have info about repos from where components are taken, so you have information about the version you use and the source from where you can detect the latest version.
> 
> It is too obvious answer for that question so I'm not sure it was the actual point for that question.

Your current proposal is something along the lines of:

'_Verify that the risk of using vulnerable 3rd party component (version) is minimized by using up to date or latest "known to be safe" version._'

How do you verify that? Clearly it is not possible "black box". So that takes us to whitebox. Does the tester have to run an SCA tool on their entire codebase? Does the application owner have to have an SCA tool running (which is what you would call a hidden requirement 😀)?

Without an SCA, this seems like an impossible task unless you can tell me otherwise so let's assume there is an SCA tool running.

SCA tools usually report on vulnerabilities, but that is not what you are asking, You are asking, are they running the most recent version of the library. So you now need to analyse the SCA report to try and figure out what the most recent version of each library is and whether they are using it. Now you didn't say it but do you mean most recent version or most recent supported versions? Many libraries will have a couple of supported minor or major versions so we need to clarify what we mean about that. We also need to hope that the SCA tool can provide that distinction.


My overall point here is that even verifying this requirement seems incredibly complicated and time consuming. 

Actually complying with it for any reasonably sized application would probably bankrupt the organization.

**Comment by elarlang** [comment]
> Your current proposal is something along the lines of:

> _'Verify that the risk of using vulnerable 3rd party component (version) is minimized by using up to date or latest "known to be safe" version.'_

Why the partial copy-paste to show the direction in different spotlight?
> Verify that the risk of using vulnerable 3rd party component (version) is minimized by using up to date or latest "known to be safe" version, **-add here your other options-**

What are your proposals to mitigate the defined risk?

**Comment by tghosth** [comment]
> > Your current proposal is something along the lines of:
> 
> > _'Verify that the risk of using vulnerable 3rd party component (version) is minimized by using up to date or latest "known to be safe" version.'_
> 
> Why the partial copy-paste to show the direction in different spotlight?

I don't really understand your point. That was your position and the only part I skipped was the bit where you ask other people for their options.

> > Verify that the risk of using vulnerable 3rd party component (version) is minimized by using up to date or latest "known to be safe" version, **-add here your other options-**
> 
> What are your proposals to mitigate the defined risk?

I have made several proposals above:

> _Verify that only component versions without exploitable vulnerabilities are in use._

I don't love this but it makes the verification requirement very clear, whilst leaving the implementation side a little fuzzy to allow companies to approach this in the way that works for them. I would be happy for this to be L1.

> _Verify that the application documentation defines a clear risk based remediation time frame for vulnerable 3rd party libraries vulnerabilities and that the application does not contain components which have breached these time frames._

This is a far more realistic requirement which makes verification and implementation very clear. I feel that the level of effort is out of sync with the intention of L1 but I could maybe be persuaded,

**Comment by randomstuff** [comment]
> SCA tools usually report on vulnerabilities, but that is not what you are asking, You are asking, are they running the most recent version of the library. So you now need to analyse the SCA report to try and figure out what the most recent version of each library is and whether they are using it. Now you didn't say it but do you mean most recent version or most recent supported versions? Many libraries will have a couple of supported minor or major versions so we need to clarify what we mean about that. We also need to hope that the SCA tool can provide that distinction.

NPM-based projects typically have thousands of (transitive) dependencies (eg. 1836 transitive dependencies for some frontend-project I'm looking at). `npm` has tooling which (I think) kind-of helps "updating to the latest" version. However, there is a risk of introducing breaking changes, regressions (including performance regressions) when doing that which means that you might have to add a manual test of all the project after doing that.

Many dependency frameworks do not really support "update to the latest version". If we take for example, Gradle (which is the gold standard of project manager and package management solution for Java), I don't think there is any builtin tooling which helps you "updating to the latest version". I think this is (sadly) very cumbersome to do for transitive dependencies (especially with complex builds such as composite builds). Moreover,  I think I have seen cases where updating a dependency A pulls a new versions of a transitive dependency B which turns out to introduce a breaking change compared to the version of B you are directly using (or using through another transitive dependency). This is not supposed to happen but this does happen. We could disregard this as "not our problem, this is a problem with the tooling and/or the ecosystem" but this would still be problematic to implement.

Summary: Dependency hell can be really messy in "big" projects pulling lots of dependencies which might make "udpate to the latest dependency" not very tractable.

**Comment by tghosth** [comment]
> > SCA tools usually report on vulnerabilities, but that is not what you are asking, You are asking, are they running the most recent version of the library. So you now need to analyse the SCA report to try and figure out what the most recent version of each library is and whether they are using it. Now you didn't say it but do you mean most recent version or most recent supported versions? Many libraries will have a couple of supported minor or major versions so we need to clarify what we mean about that. We also need to hope that the SCA tool can provide that distinction.
> 
> NPM-based projects typically have thousands of (transitive) dependencies (eg. 1836 transitive dependencies for some frontend-project I'm looking at). `npm` has tooling which (I think) kind-of helps "updating to the latest" version. However, there is a risk of introducing breaking changes, regressions (including performance regressions) when doing that which means that you might have to add a manual test of all the project after doing that.
> 
> Many dependency frameworks do not really support "update to the latest version". If we take for example, Gradle (which is the gold standard of project manager and package management solution for Java), I don't think there is any builtin tooling which helps you "updating to the latest version". I think this is (sadly) very cumbersome to do for transitive dependencies (especially with complex builds such as composite builds). Moreover, I think I have seen cases where updating a dependency A pulls a new versions of a transitive dependency B which turns out to introduce a breaking change compared to the version of B you are directly using (or using through another transitive dependency). This is not supposed to happen but this does happen. We could disregard this as "not our problem, this is a problem with the tooling and/or the ecosystem" but this would still be problematic to implement.
> 
> Summary: Dependency hell can be really messy in "big" projects pulling lots of dependencies which might make "udpate to the latest dependency" not very tractable.

Yes I think think this is at least part of my point :)

**Comment by tghosth** [comment]
Suggested wording based on discussing with Elar:

Doc requirement (need to check if anything like this previously existed): 
Verify that the application documentation defines risk based remediation time frames for 3rd party component versions with vulnerabilities and for updating libraries in general, to minimize the risk from these components.

Implementation requirement:
Verify that the application only contains components which have not breached the documented update and remediation time frames.

(both L1)

@elarlang to review and confirm

**Comment by elarlang** [comment]
ok for me

**Comment by tghosth** [comment]
> Suggested wording based on discussing with Elar:
> 
> Doc requirement (need to check if anything like this previously existed): Verify that the application documentation defines risk based remediation time frames for 3rd party component versions with vulnerabilities and for updating libraries in general, to minimize the risk from these components.
> 
> Implementation requirement: Verify that the application only contains components which have not breached the documented update and remediation time frames.
> 
> (both L1)
> 
> @elarlang to review and confirm

@randomstuff what do you think about this?

**Comment by randomstuff** [comment]
It somewhat feels weird not having a stronger requirement about not using dependencies with exploitable impactful vulnerabilities.

I'd want to have (in addition) something such as:

> Verify that all known vulnerabilities in dependencies are not exploitable or have a acceptable impact on the application.

but maybe it does not make sense to have this in addition to

> Verify that the application only contains components which have not breached the documented update and remediation time frames.

I guess the idea behind the latter is to properly take into account the time-frame between vulnerability discovery.

**Comment by tghosth** [comment]
Yeah from my perspective the key thing is to have clearly defined time scales and that should take into account exploitablity/impact. I am going to open a PR>

**Comment by tghosth** [comment]
Opened #2489 to resolve this
